### PR TITLE
feat(trezor): add HF15 support, BP+

### PR DESCRIPTION
**Pull request by ph4r05**
- HP15 changes implemented to the Trezor client in Monero wallet
- BP+ support added for Trezor
- old Trezor firmware version support removed, code cleanup

Trezor firmware changes: https://github.com/trezor/trezor-firmware/pull/2232

### CODE REVIEW

**Comment by ph4r05** [comment]
Pinging @sethforprivacy that we have all code ready in PRs to support HF15

**Comment by sethforprivacy** [comment]
> Pinging @sethforprivacy that we have all code ready in PRs to support HF15

Wow, that was quick, thank you! Will forward this on and update the checklist.

**Comment by ph4r05** [comment]
Just pinging that Trezor PRs were merged
- https://github.com/trezor/trezor-firmware/pull/2232
- https://github.com/trezor/trezor-firmware/pull/2219

**Comment by ph4r05** [comment]
Just pinging that Trezor PRs were merged
- https://github.com/trezor/trezor-firmware/pull/2232
- https://github.com/trezor/trezor-firmware/pull/2219

**Comment by sethforprivacy** [comment]
Thanks, @ph4r05, I pinged the Monero dev channel that this should be ready for review etc. yesterday, and will be sure we get it in ASAP.

Thank you for all of the hard work on this!

**Comment by kayabaNerve** [comment]
Just to confirm, does the have the inv-8 reversion? https://github.com/monero-project/monero/pull/8277 (only merged 17 days ago)

I remember hearing that concern. The only branch with a commit since then (besides a version bump on the same day, present in this branch) is the "hackathon results" which seems to be cleaning up.

Any test on the current master which confirms TXs work would be sufficient to confirm :)

**Comment by ph4r05** [comment]
@kayabaNerve thats a good point, let me check

**Comment by ph4r05** [comment]
@kayabaNerve you were right, I've incorporated the #8277 change. Trezor firmware needs the change as well, I will create the PR now.

**Comment by ph4r05** [comment]
No problem @kayabaNerve! I am glad we caught it before release! :)

**Comment by selsta** [comment]
We will have a second release before the HF that will include this patch.

**Comment by selsta** [comment]
View tags: https://github.com/monero-project/monero/pull/8061

**Comment by ph4r05** [comment]
There is definitely no view tags support added in Trezor. I will check it out this week and figure out required changes.

**Comment by selsta** [comment]
I asked @j-berman and he said

```
16:19 <jberman[m]> AFAICT trezor should be fine because they lean on the default `generate_output_ephemeral_keys` which handles view tags, but I'd need to test to know for sure and I don't have a trezor
```

but he only took a quick look himself. I think the easiest way would be to make a transaction on testnet, or change the fork height locally and try to construct a transaction.

**Comment by ph4r05** [comment]
@selsta are HF15 changes enforced or is it still possible to send HF14 transactions to the network (normal Bulletproofs)?
If HF15 rules are enforced, Trezor won't be able to sign new transactions until new trezor-firmware update.

**Comment by kayabaNerve** [comment]
I believe there will be a 24 hour interim period where both transaction types are valid. After those 24 hours, only the new ones will be.

**Comment by selsta** [comment]
@matejcik Would it be possible to split the firmware update in two? One for monero and one for the other planned changes? If not we will have to ask the monero community if they prefer to delay the hardfork or if they are okay with leaving Trezor users unable to transact for about 2 weeks. Both are not ideal :/

**Comment by matejcik** [comment]
@selsta thing is, there are no other planned changes right now .)
but good news is, we considered this and decided to go ahead with July release. So the new firmware should be out by July 20.

**Comment by matejcik** [comment]
Bad news everyone.
We will not be able to do a firmware release this month. The originally announced date, August 17, is back in the game.
I've seen that you pushed back HF activation to August 13, so that is only four days of service disruption.

**Comment by selsta** [comment]
Thank you for the update, four days shouldn't be too bad.

@ph4r05 Is there a way for me to install the firmware before the update is out? A nightly firmware build or something similar? I want to do some additional testing on testnet.

**Comment by matejcik** [comment]
@selsta you should be able to pick a firmware image [from our CI](https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/pipelines)
pick a pipeline, the task you're looking for is `core fw regular build`, then browse the artifacts.
[here is latest master build](https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/jobs/2692679661/artifacts/browse)

**Comment by selsta** [comment]
Did some testing with the firmware image from CI, tx creation and receiving worked well on testnet.

**Comment by selsta** [comment]
@ph4r05 please also open against release-v0.18

File: src/device_trezor/device_trezor.cpp [hunk 1]
```diff
@@ -697,13 +693,13 @@ namespace trezor {
     unsigned device_trezor::client_version()
     {
       auto trezor_version = get_version();
-      if (trezor_version <= pack_version(2, 0, 10)){
-        throw exc::TrezorException("Trezor firmware 2.0.10 and lower are not supported. Please update.");
+      if (trezor_version < pack_version(2, 4, 3)){
+        throw exc::TrezorException("Minimal Trezor firmware version is 2.4.3. Please update.");
```
Comment (j-berman) [inline src/device_trezor/device_trezor.cpp:697]: From [these comments](https://github.com/trezor/trezor-firmware/pull/2232#issuecomment-1112419292):

> Another thing, I hard-coded to the monero wallet that firmware version 2.5.1 supports HF15

> We are releasing 2.5.1 which does not include this PR. Please bump to 2.5.2 in the monero PR.

I'm assuming only Trezor v2.5.2 will support bp+ and view tags. Wouldn't that mean the minimal Trezor version should be v2.5.2? Is the minimum set at v2.4.3 to support tx construction pre-fork?

**Review by j-berman — APPROVED** [review]
Changes LGTM. I see the majority of this PR's changes get rid of dead code for now unusable versions, thank you :)

The one thing I didn't fully wrap my around is the minimal versioning setting, but I figure you've got it how you want it

File: src/device_trezor/device_trezor.cpp (hunk 1 above)
Comment (j-berman) [reply src/device_trezor/device_trezor.cpp:697]: Nevermind, I see. Once this ships, Trezors v2.4.3 and up will be able to use those wallets until the fork

Comment (ph4r05) [reply src/device_trezor/device_trezor.cpp:697]: Thanks for the review @j-berman! That was the idea, so old Trezors could still use the wallet even if they did not update yet. We can later enforce minimal version 2.5.2.

**Comment by ph4r05** [comment]
> @ph4r05 please also open against release-v0.18

Done: https://github.com/monero-project/monero/pull/8299  :)
thanks!

**Pull request merged**